### PR TITLE
Enable SMA Blackbox Tests for Metrics and Operations.

### LINKED ISSUE #348: Enable SMA Blackbox Tests for Metrics and Operations
PRs for https://github.com/edgexfoundry/edgex-go/issues/2008 and https://github.com/edgexfoundry/edgex-go/issues/2009 were recently merged.

Please uncomment the related blackbox tests added in https://github.com/edgexfoundry/blackbox-testing/issues/272 (PR https://github.com/edgexfoundry/blackbox-testing/pull/277) and verify they pass given the above-mentioned changes to the SMA.

### PR DESCRIPTION
**Pull request by akramtexas**
Fix #348

### CODE REVIEW

File: bin/postman-test/collections/system-management.postman_collection.json [hunk 1]
```diff
@@ -273,19 +281,23 @@
           "raw": ""
         },
         "url": {
-          "raw": "{{systemManagementUrl}}/api/v1/ping",
+          "raw": "{{systemManagementUrl}}/api/v1/metrics/device-virtual,edgex-core-command,edgex-core-data,edgex-core-metadata,edgex-support-logging,edgex-support-notifications,edgex-support-scheduler",
+          "protocol": "http",
           "host": [
-            "{{systemManagementUrl}}"
+            "localhost"
```
Comment (cherrycl) [inline bin/postman-test/collections/system-management.postman_collection.json]: The host should set to variable {{systemManagementUrl}}.

**Review by cherrycl — COMMENTED** [review]
I didn't get any response when running restart_all_services testcase. Is this an issue for service?

**Comment by akramtexas** [comment]
Rebased.

File: bin/postman-test/collections/system-management.postman_collection.json (hunk 1 above)
Comment (akramtexas) [reply bin/postman-test/collections/system-management.postman_collection.json]: I have addressed this and pushed the update 👍

**Comment by akramtexas** [comment]
> I didn't get any response when running restart_all_services testcase. Is this an issue for service?

I am investigating.

**Comment by AnthonyMBonafide** [comment]
> I didn't get any response when running restart_all_services testcase. Is this an issue for service?

@akramtexas, did we look into this issue?

**Comment by akramtexas** [comment]
> > I didn't get any response when running restart_all_services testcase. Is this an issue for service?
> 
> @akramtexas, did we look into this issue?

@AnthonyMBonafide I am looking into this right now 👍

**Comment by AnthonyMBonafide** [comment]
The changes look good. However, I am going to lean on @cherrycl for a verification since my setup for running the tests is a little different than the default configuration.

**Comment by akramtexas** [comment]
> The changes look good. However, I am going to lean on @cherrycl for a verification since my setup for running the tests is a little different than the default configuration.

@AnthonyMBonafide Excellent, and thanks for reviewing. Sounds good 👍

**Comment by akramtexas** [comment]
> I didn't get any response when running restart_all_services testcase. Is this an issue for service?

@cherrycl All test cases (including `restart_all_services`) pass.
- Start Dockerized services, beginning with first the command `docker-compose down`
- Then did `docker-compose up -d`
- Finally, ran all black box test cases (per the usual `bin/postman-test/collections/system-management.postman_collection.json`)
```
...
...
Root restart_all_services
  POST http://localhost:48090/api/v1/operation [200 OK, 571B, 4.6s]
  Pass  Status code is 200
  Pass  Response time is less than 90000ms
  Pass  Content-Type is application/json

Root operation restart
  POST http://localhost:48090/api/v1/operation [200 OK, 210B, 3.3s]
  Pass  Status code is 200
  Pass  Response time is less than 90000ms
  Pass  Content-Type is application/json
...
...
-------------------------------------------------
|                         | executed |   failed |
--------------------------+----------+-----------
|              iterations |        1 |        0 |
--------------------------+----------+-----------
|                requests |       11 |        0 |
--------------------------+----------+-----------
|            test-scripts |       11 |        0 |
--------------------------+----------+-----------
|      prerequest-scripts |        0 |        0 |
--------------------------+----------+-----------
|              assertions |       33 |        0 |
-------------------------------------------------
| total run duration: 16.1s                     |
-------------------------------------------------
| total data received: 10.56KB (approx)         |
-------------------------------------------------
| average response time: 1434ms                 |
-------------------------------------------------
Process finished with exit code 0
```

**Comment by akramtexas** [comment]
@cherrycl Cordial reminder that this `PR` is ready for your re-review 👍

**Comment by cherrycl** [comment]
@akramtexas I still didn't get any response when running restart_all_services testcase.
```
→ restart_all_services
  POST http://edgex-sys-mgmt-agent:48090/api/v1/operation [errored]
     socket hang up
  3. Status code is 200
  4. Response time is less than 90000ms
  5. Content-Type is application/json

→ operation restart
  POST http://edgex-sys-mgmt-agent:48090/api/v1/operation [200 OK, 210B, 3.6s]
  ✓  Status code is 200
  ✓  Response time is less than 90000ms
  ✓  Content-Type is application/json
...

┌─────────────────────────┬────────────────────┬────────────────────┐
│                         │           executed │             failed │
├─────────────────────────┼────────────────────┼────────────────────┤
│              iterations │                  1 │                  0 │
├─────────────────────────┼────────────────────┼────────────────────┤
│                requests │                 11 │                  1 │
├─────────────────────────┼────────────────────┼────────────────────┤
│            test-scripts │                 11 │                  0 │
├─────────────────────────┼────────────────────┼────────────────────┤
│      prerequest-scripts │                  0 │                  0 │
├─────────────────────────┼────────────────────┼────────────────────┤
│              assertions │                 33 │                  4 │
├─────────────────────────┴────────────────────┴────────────────────┤
│ total run duration: 27.5s                                         │
├───────────────────────────────────────────────────────────────────┤
│ total data received: 10.12KB (approx)                             │
├───────────────────────────────────────────────────────────────────┤
│ average response time: 1556ms [min: 7ms, max: 4.5s, s.d.: 1664ms] │
└───────────────────────────────────────────────────────────────────┘

  #  failure         detail                                                                                               
                                                                                                                          
 1.  AssertionError  Response time is less than 600ms                                                                     
                     expected 723 to be below 600                                                                         
                     at assertion:1 in test-script                                                                        
                     inside "config"                                                                                      
                                                                                                                          
 2.  Error                                                                                                                
                     socket hang up                                                                                       
                     at request                                                                                           
                     inside "restart_all_services"                                                                        
                                                                                                                          
 3.  AssertionError  Status code is 200                                                                                   
                     expected { Object (id, _details, ...) } to have property 'code'                                      
                     at assertion:0 in test-script                                                                        
                     inside "restart_all_services"                                                                        
                                                                                                                          
 4.  AssertionError  Response time is less than 90000ms                                                                   
                     expected undefined to be a number or a date                                                          
                     at assertion:1 in test-script                                                                        
                     inside "restart_all_services"                                                                        
                                                                                                                          
 5.  AssertionError  Content-Type is application/json                                                                     
                     object tested must be an array, a map, an object, a set, a string, or a weakset, but undefined given 
                     at assertion:2 in test-script                                                                        
                     inside "restart_all_services"                                                                        
Info:System-Management Test Completed.
```

**Comment by akramtexas** [comment]
Rebased.

**Comment by cherrycl** [comment]
@akramtexas  I can send the API successfully when the services only contain 2 or below services. When the services contain more than 2 services, no response returned.

If this is not test script issue, this PR is good to me.
```
API : {{systemManagementUrl}}/api/v1/operation
Request body : 
{"action":"restart", "services":["edgex-support-logging","edgex-support-notifications",] }

Response body:
[
    {
        "Success": true,
        "executor": "docker",
        "operation": "restart",
        "service": "edgex-support-notifications"
    },
    {
        "Success": true,
        "executor": "docker",
        "operation": "restart",
        "service": "edgex-support-logging"
    }
]
```

**Comment by michaelestrin** [comment]
> I can send the API successfully when the services only contain 2 or below services. When the services contain more than 2 services, no response returned.

I wonder if this is a Docker-specific limitation.  Operations are handled via executor.  The executor delegates to the Docker executable.  For a single request containing multiple operations, each operation results in a concurrent call to the executor (which in turn results in a concurrent call to the Docker executable).  I can envision the above-mentioned issue if the Docker executable is only handling one service operation at a time (in this case restart).  

You'll need to confirm this behavior @akramtexas.

**Comment by akramtexas** [comment]
> > I can send the API successfully when the services only contain 2 or below services. When the services contain more than 2 services, no response returned.
> 
> I wonder if this is a Docker-specific limitation. Operations are handled via executor. The executor delegates to the Docker executable. For a single request containing multiple operations, each operation results in a concurrent call to the executor (which in turn results in a concurrent call to the Docker executable). I can envision the above-mentioned issue if the Docker executable is only handling one service operation at a time (in this case restart).
> 
> You'll need to confirm this behavior @akramtexas.

@michaelestrin I will investigate 👍

**Comment by akramtexas** [comment]
- I need to investigate some more... My **findings so far** are as follows:
- I launched all EdgeX services (except the SMA, which I launched locally) as Dockerized containers.
Added the following two `print` statements (in hindsight, I should make these `logging` statements and get the timestamp info as well!)...
1. `fmt.Println(fmt.Sprintf("In Do()...(serviceName, operation) -> %s %s ", serviceName, operation))`
2. `fmt.Println(fmt.Sprintf("In delegateToExecutor()...(serviceName, operation) -> %s %s ", serviceName, operation))`
...respectively, as follows

```
// Do concurrently delegates a start/stop/restart operation request to the configuration-defined executor.
func (e operations) Do(services []string, operation string) []interface{} {
  var closures []concurrent.Closure
  for index := range services {
    closures = append(
      closures,
      func(serviceName string) concurrent.Closure {
        return func() interface{} {
          fmt.Println(fmt.Sprintf("In Do()...(serviceName, operation) -> %s %s ", serviceName, operation))
          return e.delegateToExecutor(serviceName, operation)
        }
      }(services[index]),
    )
  }
  return concurrent.ExecuteAndAggregateResults(closures)
}
```
...and

```
// delegateToExecutor wraps executor execution and handles error response creation when necessary.
func (e operations) delegateToExecutor(serviceName, operation string) interface{} {
  r, err := e.executor(e.executorPath, serviceName, operation)
  fmt.Println(fmt.Sprintf("In delegateToExecutor()...(serviceName, operation) -> %s %s ", serviceName, operation))
  if err != nil {
    return system.Failure(serviceName, operation, UnknownExecutorType, err.Error())
  }
  return response.Process(r, e.loggingClient)
}
```

- Invoked "restart" operation as follows to get insight into the sequencing of concurrent invocations (to identify patterns in the order, _**if any**_) for a single request containing multiple operations, where each operation results in a concurrent call to the executor (which in turn results in a concurrent call to the Docker executable):

- **[1st of two invocations]**
Endpoint: http://localhost:48090/api/v1/operation
Payload:
```
{
   "action":"restart",
   "services":["edgex-core-data","edgex-core-metadata","edgex-core-command"],
   "params":[
    "graceful"
    ]
}
```

```
Response:
[
    {
        "Success": true,
        "executor": "docker",
        "operation": "restart",
        "service": "edgex-core-data"
    },
    {
        "Success": true,
        "executor": "docker",
        "operation": "restart",
        "service": "edgex-core-command"
    },
    {
        "Success": true,
        "executor": "docker",
        "operation": "restart",
        "service": "edgex-core-metadata"
    }
]
```

Logging from instrumented code:
```
In Do()...(serviceName, operation) -> edgex-core-data restart 
In Do()...(serviceName, operation) -> edgex-core-command restart 
In Do()...(serviceName, operation) -> edgex-core-metadata restart 
In delegateToExecutor()...(serviceName, operation) -> edgex-core-data restart 
In delegateToExecutor()...(serviceName, operation) -> edgex-core-command restart 
In delegateToExecutor()...(serviceName, operation) -> edgex-core-metadata restart 
```

- **[2nd of two invocations]**
Endpoint: http://localhost:48090/api/v1/operation
Payload:
```
{
   "action":"restart",
   "services":["edgex-core-data","edgex-core-metadata","edgex-core-command","edgex-support-notifications","edgex-support-scheduler"],
   "params":[
    "graceful"
    ]
}
```

Response:
```
[
    {
        "Success": true,
        "executor": "docker",
        "operation": "restart",
        "service": "edgex-support-notifications"
    },
    {
        "Success": true,
        "executor": "docker",
        "operation": "restart",
        "service": "edgex-core-metadata"
    },
    {
        "Success": true,
        "executor": "docker",
        "operation": "restart",
        "service": "edgex-core-command"
    },
    {
        "Success": true,
        "executor": "docker",
        "operation": "restart",
        "service": "edgex-support-scheduler"
    },
    {
        "Success": true,
        "executor": "docker",
        "operation": "restart",
        "service": "edgex-core-data"
    }
]
```

Logging from instrumented code:
```
In Do()...(serviceName, operation) -> edgex-core-data restart 
In Do()...(serviceName, operation) -> edgex-core-command restart 
In Do()...(serviceName, operation) -> edgex-support-scheduler restart 
In Do()...(serviceName, operation) -> edgex-support-notifications restart 
In Do()...(serviceName, operation) -> edgex-core-metadata restart 
In delegateToExecutor()...(serviceName, operation) -> edgex-support-notifications restart 
In delegateToExecutor()...(serviceName, operation) -> edgex-core-metadata restart 
In delegateToExecutor()...(serviceName, operation) -> edgex-core-command restart 
In delegateToExecutor()...(serviceName, operation) -> edgex-support-scheduler restart 
In delegateToExecutor()...(serviceName, operation) -> edgex-core-data restart 
```

**Comment by akramtexas** [comment]
Rebased.

**Comment by michaelestrin** [comment]
Based on my conversation with @akramtexas, it appears as though multiple concurrent executions of the Docker executable for restart are actually handled concurrently by the Docker executable (i.e. all of the "before calling the executor" timestamps are roughly the same and all of the "after calling the executor" timestamps are roughly the same).

Further, it appears the Docker executable returns immediately when asking to start or stop a service.

Restart, however, takes longer (on the order of 5 seconds in Akram's testing) -- perhaps because Docker executable is actually waiting for a service to come to a complete stop before attempting to start it.

**Comment by akramtexas** [comment]
> Based on my conversation with @akramtexas, it appears as though multiple concurrent executions of the Docker executable for restart are actually handled concurrently by the Docker executable (i.e. all of the "before calling the executor" timestamps are roughly the same and all of the "after calling the executor" timestamps are roughly the same).
> 
> Further, it appears the Docker executable returns immediately when asking to start or stop a service.
> 
> Restart, however, takes longer (on the order of 5 seconds in Akram's testing) -- perhaps because Docker executable is actually waiting for a service to come to a complete stop before attempting to start it.

- In following up, I have recreated (i.e. quantified as best as possible) the behavior of the system when the BB tests are run back-to-back after a clean start of all services.
- I have created #372 (`The SMA operation "restart" (via BB scripts) fails on second and consecutive attempts`) and captured those details there.

**Comment by akramtexas** [comment]
> @akramtexas I can send the API successfully when the services only contain 2 or below services. When the services contain more than 2 services, no response returned.
> 
> If this is not test script issue, this PR is good to me.
> 
> ```
> API : {{systemManagementUrl}}/api/v1/operation
> Request body : 
> {"action":"restart", "services":["edgex-support-logging","edgex-support-notifications",] }
> 
> Response body:
> [
>     {
>         "Success": true,
>         "executor": "docker",
>         "operation": "restart",
>         "service": "edgex-support-notifications"
>     },
>     {
>         "Success": true,
>         "executor": "docker",
>         "operation": "restart",
>         "service": "edgex-support-logging"
>     }
> ]
> ```

@cherrycl Could you please provide details regarding what kind of response you are receiving? Is it an HTTP 400 status, or an empty body, etc.? (You had mentioned "no response returned".)
Also, could you please clarify whether your comment applies to the following action only, or others as well?: `restart`

**Comment by cherrycl** [comment]
@akramtexas I didn't get any return code when the services contain more than 2 services.
Only "Restart" action no response, "Stop" and "Start" works fine.
```
curl -X POST http://localhost:48090/api/v1/operation -H "Content-Type=application/json" -d '{"action":"restart", "services":["edgex-support-logging","edgex-support-notifications"] }'
[{"Success":true,"executor":"docker","operation":"restart","service":"edgex-support-notifications"},{"Success":true,"executor":"docker","operation":"restart","service":"edgex-support-logging"}]

curl -X POST http://localhost:48090/api/v1/operation -H "Content-Type=application/json" -d '{"action":"restart", "services":["edgex-support-logging","edgex-support-notifications","edgex-core-metadata", "edgex-core-data", "edgex-core-command"] }'
curl: (52) Empty reply from server
```

**Comment by michaelestrin** [comment]
@cherrycl I think curl may be timing out before a response is returned.  Approximately how many seconds does it take between you sending ```curl -X POST http://localhost:48090/api/v1/operation -H "Content-Type=application/json" -d '{"action":"restart", "services":["edgex-support-logging","edgex-support-notifications","edgex-core-metadata", "edgex-core-data", "edgex-core-command"] }'``` and curl returning the ```curl: (52) Empty reply from server``` response?

**Comment by akramtexas** [comment]
While we wait to hear back from @cherrycl on the time-lag he is seeing for that `cURL` command, I'll add that I'm consistently seeing a lag of 5 to 6 seconds. For example, per this recent logging with timestamps:
```
level=INFO ts=2020-01-07T14:19:51.826478Z app=edgex-sys-mgmt-agent source=operation.go:67 msg="In Do() -> edgex-core-metadata restart "
level=INFO ts=2020-01-07T14:19:51.826506Z app=edgex-sys-mgmt-agent source=operation.go:67 msg="In Do() -> edgex-support-notifications restart "
level=INFO ts=2020-01-07T14:19:51.826488Z app=edgex-sys-mgmt-agent source=operation.go:67 msg="In Do() -> edgex-support-logging restart "
level=INFO ts=2020-01-07T14:19:51.826516Z app=edgex-sys-mgmt-agent source=operation.go:67 msg="In Do() -> edgex-core-data restart "
level=INFO ts=2020-01-07T14:19:51.826532Z app=edgex-sys-mgmt-agent source=operation.go:67 msg="In Do() -> edgex-core-command restart "
level=INFO ts=2020-01-07T14:19:56.940057Z app=edgex-sys-mgmt-agent source=operation.go:51 msg="In delegateToExecutor() -> edgex-core-metadata restart "
level=INFO ts=2020-01-07T14:19:57.150403Z app=edgex-sys-mgmt-agent source=operation.go:51 msg="In delegateToExecutor() -> edgex-core-command restart "
level=INFO ts=2020-01-07T14:19:57.238099Z app=edgex-sys-mgmt-agent source=operation.go:51 msg="In delegateToExecutor() -> edgex-core-data restart "
level=INFO ts=2020-01-07T14:19:57.354371Z app=edgex-sys-mgmt-agent source=operation.go:51 msg="In delegateToExecutor() -> edgex-support-logging restart "
level=INFO ts=2020-01-07T14:19:57.398518Z app=edgex-sys-mgmt-agent source=operation.go:51 msg="In delegateToExecutor() -> edgex-support-notifications restart "
```
The response I get from that cURL is the same as yours:
```
curl -X POST http://localhost:48090/api/v1/operation -H "Content-Type=application/json" -d '{"action":"restart", "services":["edgex-support-logging","edgex-support-notifications","edgex-core-metadata", "edgex-core-data", "edgex-core-command"] }'
curl: (52) Empty reply from server
```

**Comment by akramtexas** [comment]
- @cherrycl, as I mentioned in [an earlier comment](https://github.com/edgexfoundry/blackbox-testing/pull/362#issuecomment-571317885) (in this PR), I run into the same issue that you do, but that is _only_ when I run the SMA BB tests _consecutively_ after an initial clean launch of Dockerized services. That is to say, the first time I run the SMA BB tests (after an initial clean launch of Dockerized services), all black box test cases pass, including `restart_all_services`).
- In fact, I have captured the above in a separate issue (#372), which contains the steps to reproduce the scenario.
- Can you please go ahead and reproduce the scenario as well? This way, we can have confirmation that we are seeing the same (consistent) behavior: The first time I run the SMA BB tests, after an initial clean launch of Dockerized services, all black box test cases pass, including `restart_all_services`). Thanks.
- Following your confirmation regarding above, can you please approve and release this PR, with the understanding that I am addressing the issue captured in #372?

**Review by cherrycl — APPROVED** [review]
Approve this PR and will follow up the restart error on issue #372.

**Review by michaelestrin — APPROVED** [review]
Based on Cherry's approval.

**Comment by akramtexas** [comment]
> Approve this PR and will follow up the restart error on issue #372.

Thank you @cherrycl - Yes, I'm actively working on issue #372 👍

**Pull request merged**